internal/clientconn/connmetrics/conn_metrics.go

```diff
@@ -0,0 +1,183 @@
+
+type BasicCommandMetrics struct {
+	Failed int64
+	Total  int64
+}
+
+type UpdateCommandMetrics struct {
+	ArrayFilters int64
+	Failed       int64
+	Pipeline     int64
+	Total        int64
+}
+
+// newConnMetrics creates new conn metrics.
+//
+// The cmds is the list of all expected commands that could be measured.
+// After providing them, they will be set with the zero values.
+func newConnMetrics(cmds []string) *ConnMetrics {
+	return &ConnMetrics{
+		Requests: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "requests_total",
+				Help:      "Total number of requests.",
+			},
+			[]string{"opcode", "command"},
+		),
+		Responses: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "responses_total",
+				Help:      "Total number of responses.",
+			},
+			[]string{"opcode", "command", "result"},
+		),
+		AggregationStages: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "aggregation_stages_total",
+				Help:      "Total number of aggregation pipeline stages.",
+			},
+			[]string{"command", "stage"},
+		),
+		cmds: cmds,
+	}
+}
+
+// Describe implements prometheus.Collector.
+func (cm *ConnMetrics) Describe(ch chan<- *prometheus.Desc) {
+	cm.Requests.Describe(ch)
+	cm.Responses.Describe(ch)
+	cm.AggregationStages.Describe(ch)
+}
+
+// Collect implements prometheus.Collector.
+func (cm *ConnMetrics) Collect(ch chan<- prometheus.Metric) {
+	cm.Requests.Collect(ch)
+	cm.Responses.Collect(ch)
+	cm.AggregationStages.Collect(ch)
+}
+
+func (cm *ConnMetrics) GetResponses() map[string]CommandMetrics {
+	res := map[string]CommandMetrics{}
+
+	// initialize commands in the map to show zero values in the metrics output
+	for _, cmd := range cm.cmds {
+		// update related operators have more fields in the output
+		switch cmd {
+		case "update", "clusterUpdate", "findAndModify":
+			res[cmd] = UpdateCommandMetrics{}
+		default:
+			res[cmd] = BasicCommandMetrics{}
+		}
+	}
+
+	ch := make(chan prometheus.Metric)
+
+	go func() {
+		cm.Responses.Collect(ch)
+		cm.AggregationStages.Collect(ch)
+		close(ch)
+	}()
+
+	for m := range ch {
+		var content dto.Metric
+		must.NoError(m.Write(&content))
+
+		var stage int
+
+		var command, opcode, result string
+
+		for _, label := range content.GetLabel() {
+			switch label.GetName() {
+			case "command":
+				command = label.GetValue()
+			case "opcode":
+				opcode = label.GetValue()
+			case "result":
+				result = label.GetValue()
+			case "stage":
+				stage = must.NotFail(strconv.Atoi(label.GetValue()))
+			default:
+				panic(fmt.Sprintf("%s is not a valid label. Allowed: [command, opcode, result, stage]", label.GetName()))
+			}
+		}
+
+		if opcode != "OP_MSG" {
+			continue
+		}
+
+		value := int64(content.GetCounter().GetValue())
+
+		cm := res[command]
```

Comment:
You use `cm` as the shorthand for `ConnMetrics` below. If this `cm` stands for `command` better replace it with `cmd`.

Reply:
`cmdMetric` will be good?